solr/core/src/test/org/apache/solr/request/macro/TestMacroExpander.java

```diff
@@ -143,15 +142,31 @@ public void testMapExprExpandOn() {
     String oldVal = System.getProperty("StreamingExpressionMacros","false");
     System.setProperty("StreamingExpressionMacros", "true");
     try {
-      @SuppressWarnings({"rawtypes"})
-      Map expanded = MacroExpander.expand(request);
-      assertEquals("zero", ((String[])expanded.get("fq"))[0]);
-      assertEquals("one", ((String[])expanded.get("fq"))[1]);
-      assertEquals("two", ((String[]) expanded.get("fq"))[2]);
-      assertEquals("three", ((String[]) expanded.get("fq"))[3]);
-      assertEquals("one", ((String[])expanded.get("expr"))[0]);
+      Map<String, String[]> expanded = MacroExpander.expand(request);
+      assertEquals("zero", expanded.get("fq")[0]);
+      assertEquals("one", expanded.get("fq")[1]);
+      assertEquals("two", expanded.get("fq")[2]);
+      assertEquals("three", expanded.get("fq")[3]);
+      assertEquals("one", expanded.get("expr")[0]);
     } finally {
       System.setProperty("StreamingExpressionMacros", oldVal);
     }
   }
+
+  @Test
+  public void testUnbalanced() { // SOLR-13181
+    final MacroExpander meSkipOnMissingParams = new MacroExpander(Collections.emptyMap());
+    final MacroExpander meFailOnMissingParams = new MacroExpander(Collections.emptyMap(), true);
+    assertEquals("${noClose", meSkipOnMissingParams.expand("${noClose"));
```

Comment:
How about adding (say) `${goodAnswer} ${noClose` as another case and it turning into `42 ${noClose` if there is a `goodAnswer=42` mapping i.e. to demonstrate that absence of closing curly bracket partially affects expansion (as opposed to no expansion at all happening)?

Reply:
I'm not 100% clear what you suggest; let me try to guess.  Let's say the parameter map has exactly one param == goodAnswer with value 42.
````
    assertEquals("42", meSkipOnMissingParams.expand("${goodAnswer}"));
    assertEquals("42", meFailOnMissingParams.expand("${goodAnswer}"));
````
Is that what you suggest?

Reply:
`${goodAnswer} ${noClose` is what I had in mind. Pushed a corresponding commit to the PR's branch, hope you don't mind.

Reply:
Ah; thanks!